### ISSUE TITLE
Help with reference

### ISSUE THREAD

**Issue by diddlywob** [tensorflow/tensorflow#87925]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

2.18.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 24.04

### Python version

3.12.3

### GCC/compiler version

gcc (Ubuntu 13.3.0-6ubuntu2~24.04) 13.3.0

### CUDA/cuDNN version

CUDA 12.0

### GPU model and memory

NVIDIA GeForce RTX 4050

### Current behavior?

I am expecting my model to run without giving any type of warnings. I know for certain that my data does not have any NaN values in it, so I am not sure why am I getting the warning I am receiving. My output should look something like the below:

2138/14403 ━━━━━━━━━━━━━━━━━━━━ 2:45 14ms/step - accuracy: 0.7590 - loss: 1.1440E0000 00:00:1740370043.647457

### Standalone code to reproduce the issue

```shell
I know the issue has to do with this part of the code. I am just not sure where.

import os
os.environ['TF_GPU_ALLOCATOR'] = 'cuda_malloc_async'
import math
import pyarrow.parquet as pq
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow.keras.regularizers import l2
from scipy.sparse import hstack, csr_matrix
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense, Dropout, BatchNormalization
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import OneHotEncoder
from itertools import combinations

# ---- Initialize Encoders & Scaler ----
onehot_encoder = OneHotEncoder(handle_unknown="ignore", sparse_output=False)
scaler = StandardScaler()

# ---- Stream Train/Test Split in Chunks ----
def stream_split_data(parquet_files, categorical_columns, numeric_columns):
    """Splits merged data into training and testing sets dynamically in chunks."""
    for chunk in stream_merged_data(parquet_files, categorical_columns, numeric_columns, merge_key):
        mask = np.random.rand(len(chunk)) < train_ratio
        yield chunk[mask], chunk[~mask] 

# ---- Fit OneHotEncoder & StandardScaler in Chunks ----
numeric_columns_for_model = [col for col in numeric_columns if col != "Result"]
for train_chunk, _ in stream_split_data(parquet_files, categorical_columns, numeric_columns):
    X_train = train_chunk.drop(columns=["Result"]) # Do not fit target variable
    scaler.partial_fit(X_train[numeric_columns_for_model])  # Fit StandardScaler

def get_all_categories(parquet_files, categorical_columns, merge_key):
    # Initialize sets to collect unique values for each categorical column.
    all_categories = {col: set() for col in categorical_columns}
    for chunk in stream_merged_data(parquet_files, categorical_columns, numeric_columns, merge_key):
        for col in categorical_columns:
            all_categories[col].update(chunk[col].unique())
    # Convert sets to sorted lists (or lists in any consistent order)
    for col in all_categories:
        all_categories[col] = sorted(list(all_categories[col]))
    return all_categories

complete_categories = get_all_categories(parquet_files, categorical_columns, merge_key)

one_hot_encoder = OneHotEncoder(
    categories=[complete_categories[col] for col in categorical_columns],
    sparse_output=False,
    handle_unknown="ignore"
)
# Fit it on a dummy DataFrame with the correct column names.
dummy_df = pd.DataFrame([["0"] * len(categorical_columns)], columns=categorical_columns)
one_hot_encoder.fit(dummy_df)


print(f"Preprocessing the data in batches...")


# ---- Generator for Streaming Batches into `tf.data.Dataset` ----
def preprocess_dense_batches(parquet_files, categorical_columns, numeric_columns, batch_size, end_idx=None):
    """Generator function for streaming chunks into tf.data.Dataset dynamically."""

    label_mapping = {0.5: 0, 0: 1, 1: 2}  

    current_idx = 0
    for train_chunk, _ in stream_split_data(parquet_files, categorical_columns, numeric_columns):
        
        y_chunk = train_chunk.pop("Result").apply(lambda x: label_mapping[x]).values.astype(int)  # Extracts y (target variable) and drops column from the rest of the data in one step

        X_chunk = np.hstack((
            one_hot_encoder.transform(train_chunk[categorical_columns]),  
            scaler.transform(train_chunk[numeric_columns_for_model])  
        )).astype(np.float32)

        # Yield batches from each chunk
        for i in range(0, len(X_chunk), batch_size):
            if end_idx is not None and current_idx >= end_idx:
                return  # Stop when we reach end index
            X_batch = X_chunk[i:i + batch_size]
            y_batch = y_chunk[i:i + batch_size]

            yield X_batch, y_batch
            current_idx += batch_size

# ---- Create `tf.data.Dataset` Using the Chunked Generator ----
train_dataset = tf.data.Dataset.from_generator(
    lambda: preprocess_dense_batches(parquet_files, categorical_columns, numeric_columns, batch_size, end_idx=int(train_ratio * total_rows)),
    output_signature=(
        tf.TensorSpec(shape=(None, one_hot_encoder.transform(pd.DataFrame([["0"] * len(categorical_columns)], columns=categorical_columns)).shape[1] + len(numeric_columns_for_model)), dtype=tf.float32),
        tf.TensorSpec(shape=(None,), dtype=tf.float32)
    )
).repeat().prefetch(buffer_size=tf.data.AUTOTUNE)

test_dataset = tf.data.Dataset.from_generator(
    lambda: preprocess_dense_batches(parquet_files, categorical_columns, numeric_columns, batch_size, end_idx=total_rows),
    output_signature=(
        tf.TensorSpec(shape=(None, one_hot_encoder.transform(pd.DataFrame([["0"] * len(categorical_columns)], columns=categorical_columns)).shape[1] + len(numeric_columns_for_model)), dtype=tf.float32),
        tf.TensorSpec(shape=(None,), dtype=tf.float32)
    )
).prefetch(buffer_size=tf.data.AUTOTUNE)

class_weight_dict = {
    0: 1.0, 
    1: 2.0,  
    2: 3.0   
}

steps_per_epoch = math.ceil((train_ratio * total_rows) / batch_size)
test_steps = math.ceil(((1 - train_ratio) * total_rows) / batch_size)

print("Datasets created dynamically.")

input_all_data = Input(shape=(one_hot_encoder.transform(pd.DataFrame([["0"] * len(categorical_columns)], columns=categorical_columns)).shape[1] + len(numeric_columns_for_model),), name='all_data_input')

print("Creating the dense layers...")

x = Dense(128, activation='relu', kernel_regularizer=l2(0.001))(input_all_data)
x = BatchNormalization()(x)
x = Dropout(0.3)(x)

x = Dense(64, activation='relu', kernel_regularizer=l2(0.001))(x)
x = BatchNormalization()(x)
x = Dropout(0.3)(x)

output = Dense(3, activation='softmax')(x)  # 3 neurons, softmax for multi-class

# Compile the model
model = Model(inputs=input_all_data, outputs=output)
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

print("\nBegin training the model.\n")

history_all_data = model.fit(
    train_dataset,                    
    validation_data=test_dataset,     
    epochs=1,                    
    steps_per_epoch=steps_per_epoch,
    validation_steps=test_steps,
    class_weight=class_weight_dict,  # Pass class weights
    verbose=1
)
```

### Relevant log output

```shell
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1740371136.641692   17996 service.cc:148] XLA service 0x7f9fc4004850 initialized for platform CUDA (this does not guarantee that XLA will be used). Devices:
I0000 00:00:1740371136.642763   17996 service.cc:156]   StreamExecutor device (0): NVIDIA GeForce RTX 4050 Laptop GPU, Compute Capability 8.9
2025-02-23 23:25:36.704967: I tensorflow/compiler/mlir/tensorflow/utils/dump_mlir_util.cc:268] disabling MLIR crash reproducer, set env var `MLIR_CRASH_REPRODUCER_DIRECTORY` to enable.
I0000 00:00:1740371136.874435   17996 cuda_dnn.cc:529] Loaded cuDNN version 90300
I0000 00:00:1740371138.351218   17996 device_compiler.h:188] Compiled cluster using XLA!  This line is logged at most once for the lifetime of the process.
 1490/14403 ━━━━━━━━━━━━━━━━━━━━ 3:33 17ms/step - accuracy: 0.5462 - loss: 2.3473E0000 00:00:1740371164.392105   17996 buffer_comparator.cc:157] Difference at 0: nan, expected 1.65142
E0000 00:00:1740371164.392245   17996 buffer_comparator.cc:157] Difference at 1: nan, expected 1.48098
E0000 00:00:1740371164.392259   17996 buffer_comparator.cc:157] Difference at 2: nan, expected 1.67361
E0000 00:00:1740371164.392283   17996 buffer_comparator.cc:157] Difference at 3: nan, expected 3.16466
E0000 00:00:1740371164.392288   17996 buffer_comparator.cc:157] Difference at 4: nan, expected 2.17621
E0000 00:00:1740371164.392296   17996 buffer_comparator.cc:157] Difference at 5: nan, expected 3.38546
E0000 00:00:1740371164.392319   17996 buffer_comparator.cc:157] Difference at 6: nan, expected 1.67085
E0000 00:00:1740371164.392324   17996 buffer_comparator.cc:157] Difference at 7: nan, expected 1.38643
E0000 00:00:1740371164.392342   17996 buffer_comparator.cc:157] Difference at 8: nan, expected 1.76836
E0000 00:00:1740371164.392347   17996 buffer_comparator.cc:157] Difference at 9: nan, expected 3.48673
2025-02-23 23:26:04.392363: E external/local_xla/xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:987] Results do not match the reference. This is likely a bug/unexpected loss of precision.
 2967/14403 ━━━━━━━━━━━━━━━━━━━━ 2:58 16ms/step - accuracy: 0.5747 - loss: 2.0899E0000 00:00:1740371185.763887   17997 buffer_comparator.cc:157] Difference at 0: nan, expected 3.13696
E0000 00:00:1740371185.764129   17997 buffer_comparator.cc:157] Difference at 1: nan, expected 3.33544
E0000 00:00:1740371185.764148   17997 buffer_comparator.cc:157] Difference at 2: nan, expected 2.92697
E0000 00:00:1740371185.764155   17997 buffer_comparator.cc:157] Difference at 3: nan, expected 2.59307
E0000 00:00:1740371185.764161   17997 buffer_comparator.cc:157] Difference at 4: nan, expected 2.88114
E0000 00:00:1740371185.764169   17997 buffer_comparator.cc:157] Difference at 5: nan, expected 2.56242
E0000 00:00:1740371185.764175   17997 buffer_comparator.cc:157] Difference at 6: nan, expected 2.62881
E0000 00:00:1740371185.764181   17997 buffer_comparator.cc:157] Difference at 7: nan, expected 2.84552
E0000 00:00:1740371185.764234   17997 buffer_comparator.cc:157] Difference at 8: nan, expected 2.49929
E0000 00:00:1740371185.764246   17997 buffer_comparator.cc:157] Difference at 9: nan, expected 2.47989
2025-02-23 23:26:25.764275: E external/local_xla/xla/service/gpu/autotuning/gemm_fusion_autotuner.cc:987] Results do not match the reference. This is likely a bug/unexpected loss of precision.
```
